### ISSUE TITLE
Inpainting modifies areas where mask == 0

### ISSUE THREAD

**Issue by rb-synth** [huggingface/diffusers#4782]
### Describe the bug

With a inpainting (also controlnet inpainting), areas outside of the mask are slightly modified. 

In the example below, we inspect the first pixel, which despite being outside of the mask `(0,0,0)`, the pixel value has changed.

### Reproduction

```python
from diffusers import AutoPipelineForInpainting
from diffusers.utils import load_image
import torch
import numpy as np

torch_device = "cuda" if torch.cuda.is_available() else "cpu"

pipe = AutoPipelineForInpainting.from_pretrained("runwayml/stable-diffusion-inpainting", torch_dtype=torch.float16, variant="fp16", use_safetensors=True)

pipe = pipe.to(torch_device)
pipe.enable_xformers_memory_efficient_attention()

img_url = "https://raw.githubusercontent.com/CompVis/latent-diffusion/main/data/inpainting_examples/overture-creations-5sI6fQgYIuo.png"
mask_url = "https://raw.githubusercontent.com/CompVis/latent-diffusion/main/data/inpainting_examples/overture-creations-5sI6fQgYIuo_mask.png"

image = load_image(img_url)
mask_image = load_image(mask_url)

prompt = "dslr photography of an empty bench, high quality"

out = pipe(
    prompt=prompt,
    image=image,
    mask_image=mask_image,
    guidance_scale=8.0,
    num_inference_steps=20,
).images[0]

print("first pixel of mask", np.array(mask_image)[0, 0])
print("first pixel of input image", np.array(image)[0, 0])
print("first pixel of output image", np.array(out)[0, 0])
```
Output
```
first pixel of mask [0 0 0]
first pixel of input image [164 184 193]
first pixel of output image [185 202 209]
```

![image](https://github.com/huggingface/diffusers/assets/135021519/2dace566-d456-4bf2-a4fc-515280be2773)
![image](https://github.com/huggingface/diffusers/assets/135021519/879d3490-a496-48e6-853b-1368c88751f5)
![dog](https://github.com/huggingface/diffusers/assets/135021519/71e6c7a8-7918-40d8-994c-971428361550)

### System Info

- `diffusers` version: 0.20.0
- Platform: Linux-6.2.0-27-generic-x86_64-with-glibc2.37
- Python version: 3.9.17
- PyTorch version (GPU?): 2.0.1+cu117 (True)
- Huggingface_hub version: 0.16.4
- Transformers version: 4.32.0
- Accelerate version: 0.22.0
- xFormers version: 0.0.21
- Using GPU in script?: <fill in>
- Using distributed or parallel set-up in script?: no

**Comment by rb-synth** [comment]
This may seem slightly pedantic, as the pixels look unchanged to the human eye -- it appears that the only difference between the input and output images is the lack of the dog. 

But in my case, the image has a uniformly grey background (128,128,128). The background of the output from SD has values of ~(127,127,127). When I blend the two together, I get a darkened halo effect around my images:

![image](https://github.com/huggingface/diffusers/assets/135021519/93c57049-0f1a-46b7-b915-d9c4f6673e56)

**Comment by xiankgx** [comment]
This is not a bug, but rather how Stable Diffusion (SD) works. In SD models, the VAE is responsible for encoding images to latents, and decoding latents images. Images are generated in the latent space which the VAE then decodes into the image space (for e.g. 64 x 64 -> to 512 x 512). In the unmasked areas (pixel values 0), the original image pixels in those areas are still encoded with the VAE. After the entire denoising diffusion process, it gets decoded back to image space. However, the decoded unmasked areas are now slightly different then what it was. In other words, the VAE intoduces some error in the reconstruction.

If you want exact pixel values in unmasked areas, then just replace the generated pixels with the original pixels.
